api/src/main/java/org/openmrs/module/patientqueueing/api/impl/PatientQueueingServiceImpl.java

```diff
@@ -12,70 +12,130 @@
+        this.dao = dao;
+    }
+
+    /**
+     * @see org.openmrs.module.patientqueueing.api.PatientQueueingService#savePatientQue(org.openmrs.module.patientqueueing.model.PatientQueue)
+     */
+    public PatientQueue savePatientQue(PatientQueue patientQueue) {
+        PatientQueue currentQueue = dao.getIncompletePatientQueue(patientQueue.getPatient(), patientQueue.getLocationTo());
+
+        if (currentQueue != null) {
+            completePatientQueue(currentQueue);
+        }
+
+        return dao.savePatientQueue(patientQueue);
+    }
+
+    /**
+     * @see org.openmrs.module.patientqueueing.api.PatientQueueingService#getPatientQueueById(java.lang.Integer)
+     */
+    public PatientQueue getPatientQueueById(Integer queueId) {
+        return dao.getPatientQueueById(queueId);
+    }
+
+    /**
+     * @see org.openmrs.module.patientqueueing.api.PatientQueueingService#getPatientQueueList(org.openmrs.Provider,
+     * java.util.Date, java.util.Date, org.openmrs.Location, org.openmrs.Location,
+     * org.openmrs.Patient, org.openmrs.module.patientqueueing.model.PatientQueue.Status)
+     */
+    public List<PatientQueue> getPatientQueueList(Provider provider, Date fromDate, Date toDate, Location locationTo, Location locationFrom, Patient patient, PatientQueue.Status status) {
+        return dao.getPatientQueueList(provider, fromDate, toDate, locationTo, locationFrom, patient, status);
+    }
+
+    /**
+     * @see org.openmrs.module.patientqueueing.api.PatientQueueingService#completePatientQueue(org.openmrs.module.patientqueueing.model.PatientQueue)
+     */
+    @Override
+    public PatientQueue completePatientQueue(PatientQueue patientQueue) {
+        patientQueue.setStatus(PatientQueue.Status.COMPLETED);
+        return dao.savePatientQueue(patientQueue);
+    }
+
+    /**
+     * @see org.openmrs.module.patientqueueing.api.PatientQueueingService#getIncompletePatientQueue(org.openmrs.Patient,
+     * org.openmrs.Location)
+     */
+    public PatientQueue getIncompletePatientQueue(Patient patient, Location locationTo) {
+
+        return dao.getIncompletePatientQueue(patient, locationTo);
+    }
+
+    /**
+     * @see org.openmrs.module.patientqueueing.api.PatientQueueingService#getMostRecentQueue(org.openmrs.Patient)
+     */
+    @Override
+    public PatientQueue getMostRecentQueue(Patient patient) {
+        return dao.getMostRecentQueue(patient);
+    }
+
+    /**
+     * @see org.openmrs.module.patientqueueing.api.PatientQueueingService#assignVisitNumber(org.openmrs.module.patientqueueing.model.PatientQueue)
+     */
+    @Override
+    public PatientQueue assignVisitNumber(PatientQueue patientQueue) {
+        Date today = new Date();
+        List<PatientQueue> patientQueueList = getPatientQueueList(null, OpenmrsUtil.firstSecondOfDay(today), OpenmrsUtil.getLastMomentOfDay(today), null, null, patientQueue.getPatient(), null);
+
+        if (!patientQueueList.isEmpty() && patientQueueList.get(0).getVisitNumber() != null) {
```

Comment:
Why consider only the first item in the list but ignore the others?

Reply:
Because they have the same visit number based on that parameters used to get it. As long as the list is not empty